src/resolvers/Promotion.js

```diff
@@ -0,0 +1,267 @@
+       * check user type is Brand owner
+       */
+      if (context.user.type !== USER_TYPE.BRANDOWNER) {
+        throw AuthenticationError("Invalid Permission");
+      }
+
+      /**
+       * disallow id
+       */
+      if (promotionInput.id !== null) {
+        throw new UserInputError("ID not allowed while creating promotion");
+      }
+
+      /**
+       * - check validity of promotion name
+       * - disallow name if the promotion is still ongoing
+       */
+      const existingPromotion = await Promotion.findOne({
+        promotionNm: promotionInput.promotionNm,
+      });
+      if (
+        existingPromotion &&
+        new Date(existingPromotion.promotionEnd) > Date.now()
+      ) {
+        throw new ApolloError("Prmotion name alread taken");
+      }
+
+      /**
+       * date range validity check:
+       * - must start at least one week from now
+       * - end date > start date
+       */
+      const validStartDt =
+        new Date(promotionInput.promotionStart) >
+        new Date(Date.now() + 1000 * 60 * 60 * 24 * 7);
+      const validEndDt =
+        new Date(promotionInput.promotionStart) <
+        new Date(promotionInput.promotionEnd);
+      if (!validStartDt || !validEndDt) {
+        throw UserInputError("Invalid date");
+      }
+
+      promotionInput.image._id = promotionInput.image.id;
+      promotionInput.promotionService.forEach((element) => {
+        element._id = element.id;
+      });
+      promotionInput.promotionBranch.forEach((element) => {
+        element._id = element.id;
+      });
+
+      let newPromotion = await Promotion.create(promotionInput);
+
+      newPromotion = promotionConverter(newPromotion);
+      newPromotion.id = newPromotion._id;
+
+      return newPromotion;
+    },
+    updatePromotion: async (root, { promotionInput }, context, info) => {
+      /**
+       * - only allow loggedin brandowner
+       * - valid id check
+       * - name validity check
+       * - date validity check
+       */
+
+      /**
+       * check user type is Brand owner
+       */
+      if (context.user.type !== USER_TYPE.BRANDOWNER) {
+        throw AuthenticationError("Invalid Permission");
+      }
+
+      const validId = mongoose.Types.ObjectId.isValid(promotionInput.id);
+      const existingPromotionById = await Promotion.findById(promotionInput.id);
+      if (!validId || existingPromotionById === null) {
+        throw new UserInputError("Invalid ID");
+      }
+
+      /**
+       * - check validity of promotion name
+       * - disallow name if the promotion is still ongoing
+       * - allow if the name is same as previous name (means no update)
+       */
+      const existingPromotion = await Promotion.findOne({
+        promotionNm: { $eq: promotionInput.promotionNm },
+      });
+      if (
+        existingPromotion &&
+        existingPromotion.promotionNm !== promotionInput.promotionNm
+      ) {
+        if (new Date(existingPromotion.promotionEnd) > Date.now()) {
```

Comment:
why check on the end date of existing promotion?
if the promotion with same name is running or not yet started then they able to have same name?

Reply:
- if end > now, means it is running. => disallow the name
- if end < now, means it is over. => allow the name